### PR TITLE
RecordIO options for reader

### PR DESCRIPTION
**Pull request by sebheitzmann**

### CODE REVIEW

File: recordio/file_reader.go [hunk 1]
```diff
@@ -350,28 +350,53 @@ func readNextV2(r *FileReader) ([]byte, error) {
 	return returnSlice, nil
 }
 
-// TODO(thomas): we have to add an option pattern here as well
```
Comment (sebheitzmann) [inline recordio/file_reader.go:353]: Hi Thomas, 

I have implemented this option pattern to be able to set the buffer size. The default one is too big for my use case.

File: simpledb/compaction.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ func backgroundCompaction(db *DB) {
 			select {
 			case <-db.compactionTickerStopChannel:
 				return nil
-			case _ = <-db.compactionTicker.C:
+			case <-db.compactionTicker.C:
```
Comment (sebheitzmann) [inline simpledb/compaction.go:32]: just a little syntax error reported by my static analyser

File: simpledb/sstable_manager.go [hunk 3]
```diff
@@ -123,10 +123,10 @@ func (s *SSTableManager) candidateTablesForCompaction(compactionMaxSizeBytes uin
 	for i := 0; i < len(s.allSSTableReaders); i++ {
 		reader := s.allSSTableReaders[i]
 		// avoid the EmptySStableReader (or empty files) and only include small enough SSTables
-		if reader.MetaData().NumRecords > 0 && reader.MetaData().TotalBytes < compactionMaxSizeBytes {
+		if reader.MetaData().TotalBytes < compactionMaxSizeBytes {
 			paths = append(paths, reader.BasePath())
 			numRecords += reader.MetaData().NumRecords
-			if i == 0 {
+			if numRecords == 0 {
```
Comment (sebheitzmann) [inline simpledb/sstable_manager.go:129]: this one is a bug if there are no record in a "delete only" sstable.

File: recordio/recordio_test.go [hunk 4]
```diff
@@ -67,7 +67,7 @@ func TestReadWriteEndToEndDirectIO(t *testing.T) {
 	require.NoError(t, err)
 
 	reader := func() ReaderI {
-		reader, err := newFileReaderWithFactory(tmpFile.Name(), DirectIOFactory{})
+		reader, err := NewFileReader(ReaderPath(tmpFile.Name()))
```
Comment (thomasjungblut) [inline recordio/recordio_test.go:70]: do you mind to make this buffer factory an option too? this path now doesn't run the DirectIO check anymore

File: simpledb/recovery.go [hunk 5]
```diff
@@ -240,7 +240,9 @@ func (db *DB) replayAndSetupWriteAheadLog() error {
 		err = executeFlush(db, memStoreFlushAction{
 			memStore: swapMemstore(db),
 		})
-
+		if err != nil {
+			log.Printf("Error Flush memstore %v", err)
+		}
```
Comment (thomasjungblut) [inline simpledb/recovery.go:245]: oof, well spotted. Do you think we should return an error here?

**Comment by thomasjungblut** [comment]
just two minor fixes, if you find some time. I'll merge right after :) Thank you!

File: recordio/recordio_test.go (hunk 4 above)
Comment (sebheitzmann) [reply recordio/recordio_test.go:70]: The DirectIOFactory is in the fonction. So i Think that this does exactly the same thing as previously. Sorry, I don't understand our comment.

File: simpledb/recovery.go (hunk 5 above)
Comment (sebheitzmann) [reply simpledb/recovery.go]: yes, modified.

File: recordio/recordio_test.go (hunk 4 above)
Comment (sebheitzmann) [reply recordio/recordio_test.go:70]: maybe you want a WithFactory() option ? Is it what you think ?

**Comment by thomasjungblut** [comment]
yes

File: recordio/recordio_test.go (hunk 4 above)
Comment (sebheitzmann) [reply recordio/recordio_test.go:70]: I added a IoFactory Interface. Is it ok for you ?

**Comment by sebheitzmann** [comment]
tests are ok. If you can merge it would be wonderfull.

File: recordio/recordio_test.go (hunk 4 above)
Comment (thomasjungblut) [reply recordio/recordio_test.go:70]: yes, but you need to pass the DirectIO factory here, it's otherwise not testing it

Comment (sebheitzmann) [reply recordio/recordio_test.go]: Done

**Review by thomasjungblut — APPROVED** [review]
lgtm, just waiting for the test to finish

**Comment by thomasjungblut** [comment]
Thank you @sebheitzmann - much appreciated :)

**Comment by sebheitzmann** [comment]
is it possible to tag a version ?

**Comment by thomasjungblut** [comment]
absolutely, give me a few minutes

**Pull request merged**